### ISSUE TITLE
BUG: Mesh parallel error

### ISSUE THREAD

**Issue by zhyxue0** [firedrakeproject/firedrake#3265]
Hello everyone. 

I try to use `project` on two different meshes. The code worked fine when running serially, but when I ran it in parallel, it prompted: `AssertionError: Whoever made mesh_B should explicitly mark mesh_A as having a compatible parallel layout.`

Here is the code:
```python
from firedrake import *


Lx = 2 
Ly = 2

Nx = 4
Ny = Nx

mesh_1 = PeriodicRectangleMesh(Nx, Ny, Lx, Ly)
V_1 = VectorFunctionSpace(mesh_1, "DG", 1)
u_1 = Function(V_1)

mesh_2 = PeriodicRectangleMesh(2*Nx, 2*Ny, Lx, Ly)
V_2 = VectorFunctionSpace(mesh_2, "DG", 1)


u_2 = project(u_1, V_2)


```

Here is the report:
```shell
Traceback (most recent call last):
  File "/home/text_1.py", line 18, in <module>
    u_2 = project(u_1, V_2)
  File "petsc4py/PETSc/Log.pyx", line 115, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
Traceback (most recent call last):
  File "/home/text_1.py", line 18, in <module>
    u_2 = project(u_1, V_2)
  File "petsc4py/PETSc/Log.pyx", line 115, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "petsc4py/PETSc/Log.pyx", line 116, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "petsc4py/PETSc/Log.pyx", line 116, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "/home/firedrake/firedrake/src/firedrake/firedrake/adjoint/projection.py", line 34, in wrapper
  File "/home/firedrake/firedrake/src/firedrake/firedrake/adjoint/projection.py", line 34, in wrapper
    output = project(*args, **kwargs)
  File "/home/firedrake/firedrake/src/firedrake/firedrake/projection.py", line 78, in project
    output = project(*args, **kwargs)
  File "/home/firedrake/firedrake/src/firedrake/firedrake/projection.py", line 78, in project
    use_slate_for_inverse=use_slate_for_inverse).project()
  File "/home/firedrake/firedrake/src/firedrake/firedrake/projection.py", line 181, in project
    use_slate_for_inverse=use_slate_for_inverse).project()
  File "/home/firedrake/firedrake/src/firedrake/firedrake/projection.py", line 181, in project
    self.apply_massinv(self.target, self.rhs)
  File "/home/firedrake/firedrake/src/firedrake/firedrake/projection.py", line 240, in rhs
    self.apply_massinv(self.target, self.rhs)
  File "/home/firedrake/firedrake/src/firedrake/firedrake/projection.py", line 240, in rhs
    self.mixed_mass.mult(u, v)
  File "/home/firedrake/firedrake/src/PyOP2/pyop2/utils.py", line 62, in __get__
    self.mixed_mass.mult(u, v)
  File "/home/firedrake/firedrake/src/PyOP2/pyop2/utils.py", line 62, in __get__
    obj.__dict__[self.__name__] = result = self.fget(obj)
  File "/home/firedrake/firedrake/src/firedrake/firedrake/projection.py", line 234, in mixed_mass
    obj.__dict__[self.__name__] = result = self.fget(obj)
  File "/home/firedrake/firedrake/src/firedrake/firedrake/projection.py", line 234, in mixed_mass
    return assemble_mixed_mass_matrix(self.source.function_space(),
  File "petsc4py/PETSc/Log.pyx", line 115, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
    return assemble_mixed_mass_matrix(self.source.function_space(),
  File "petsc4py/PETSc/Log.pyx", line 115, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "petsc4py/PETSc/Log.pyx", line 116, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "petsc4py/PETSc/Log.pyx", line 116, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "/home/firedrake/firedrake/src/firedrake/firedrake/supermeshing.py", line 97, in assemble_mixed_mass_matrix
  File "/home/firedrake/firedrake/src/firedrake/firedrake/supermeshing.py", line 97, in assemble_mixed_mass_matrix
    intersections = intersection_finder(mesh_A, mesh_B)
  File "firedrake/cython/supermeshimpl.pyx", line 132, in firedrake.cython.supermeshimpl.intersection_finder
    intersections = intersection_finder(mesh_A, mesh_B)
  File "firedrake/cython/supermeshimpl.pyx", line 132, in firedrake.cython.supermeshimpl.intersection_finder
AssertionError: Whoever made mesh_B should explicitly mark mesh_A as having a compatible parallel layout.
application called MPI_Abort(PYOP2_COMM_WORLD, 1) - process 0
AssertionError: Whoever made mesh_B should explicitly mark mesh_A as having a compatible parallel layout.
application called MPI_Abort(PYOP2_COMM_WORLD, 1) - process 1
```

**Comment by connorjward** [comment]
This is a known limitation of projection between mismatching meshes (supermeshing). It will only work if the meshes have a matching parallel decomposition which we cannot guarantee here I believe.

We do support [interpolation across meshes](https://www.firedrakeproject.org/interpolation.html#interpolation-across-meshes) into particular function spaces. This is however not quite the same as projection so may not be suitable for you.

**Comment by colinjcotter** [comment]
Those two meshes are nested so project and interpolate have the same result.

________________________________________

**Comment by connorjward** [comment]
> > This is a known limitation of projection between mismatching meshes (supermeshing). It will only work if the meshes have a matching parallel decomposition which we cannot guarantee here I believe.
> > We do support [interpolation across meshes](https://www.firedrakeproject.org/interpolation.html#interpolation-across-meshes) into particular function spaces. This is however not quite the same as projection so may not be suitable for you.
> 
> ## I tried to replace project() with interpolate() when project the results between two different meshes. However, I met the problem that : [0]PETSC ERROR: ------------------------------------------------------------------------
> [0]PETSC ERROR: Caught signal number 11 SEGV: Segmentation Violation, probably memory access out of range
> [0]PETSC ERROR: Try option -start_in_debugger or -on_error_attach_debugger
> [0]PETSC ERROR: or see https://petsc.org/release/faq/#valgrind and https://petsc.org/release/faq/
> [0]PETSC ERROR: configure using --with-debugging=yes, recompile, link, and run
> [0]PETSC ERROR: to get more information on the crash.
> [0]PETSC ERROR: Run with -malloc_debug to check if memory corruption is causing the crash.
> MPI_ABORT was invoked on rank 0 in communicator MPI_COMM_WORLD with errorcode 59.
> 
> NOTE: invoking MPI_ABORT causes Open MPI to kill all MPI processes. You may or may not see output from other processes, depending on exactly when Open MPI kills them.

This is a different error message to the one above. Can you create a new issue for your problem?
